File: app/src/main/java/com/kylemadsen/testandroid/ViewGroupController.java
package com.kylemadsen.testandroid;

import android.support.annotation.MainThread;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public class ViewGroupController {

    private final List<ViewController> viewControllers = new ArrayList<>();

    private ViewGroup viewGroup;
    private LayoutInflater layoutInflater;

    private ViewGroupController(ViewGroup viewGroup, LayoutInflater layoutInflater) {
        this.viewGroup = viewGroup;
        this.layoutInflater = layoutInflater;
    }

    @MainThread
    public static ViewGroupController onCreate(ViewGroup viewGroup) {
        LayoutInflater layoutInflater = LayoutInflater.from(viewGroup.getContext());
        return new ViewGroupController(viewGroup, layoutInflater);
    }

    @MainThread
    public <ViewControllerType extends ViewController> void attach(ViewControllerType viewController) {
        View view = layoutInflater.inflate(viewController.getLayoutId(), viewGroup, true);
        viewController.attach(view);
        viewControllers.add(viewController);
    }

    @MainThread
    public void onDestroy() {
        for (ViewController viewController : viewControllers) {
            viewController.detach();
        }
        viewControllers.clear();
        viewGroup.removeAllViews();
        viewGroup = null;
    }
}

File: core/src/main/java/com/kylemadsen/core/view/ViewGroupController.kt
package com.kylemadsen.core.view

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.MainThread

import java.util.ArrayList

class ViewGroupController private constructor(
        private var viewGroup: ViewGroup?,
        private val layoutInflater: LayoutInflater
) {

    private val viewControllers = ArrayList<ViewController>()

    @MainThread
    fun <ViewControllerType : ViewController> attach(viewController: ViewControllerType) {
        val view = layoutInflater.inflate(viewController.layoutId, viewGroup, true)
        viewController.attach(view)
        viewControllers.add(viewController)
    }

    @MainThread
    fun onDestroy() {
        for (viewController in viewControllers) {
            viewController.detach()
        }
        viewControllers.clear()
        viewGroup?.removeAllViews()
        viewGroup = null
    }

    companion object {

        @MainThread
        @JvmStatic
        fun onCreate(viewGroup: ViewGroup): ViewGroupController {
            val layoutInflater = LayoutInflater.from(viewGroup.context)
            return ViewGroupController(viewGroup, layoutInflater)
        }
    }
}
